core/iwasm/common/wasm_shared_memory.c

```diff
@@ -380,17 +390,24 @@ static bool
 map_remove_wait_info(HashMap *wait_map_, AtomicWaitInfo *wait_info,
                      void *address)
 {
+    os_mutex_lock(&wait_map_lock);
+    os_mutex_lock(&wait_info->wait_list_lock);
     if (wait_info->wait_list->len > 0) {
+        os_mutex_unlock(&wait_info->wait_list_lock);
+        os_mutex_unlock(&wait_map_lock);
         return false;
     }
+    os_mutex_unlock(&wait_info->wait_list_lock);
 
     bh_hash_map_remove(wait_map_, address, NULL, NULL);
+    os_mutex_unlock(&wait_map_lock);
     return true;
 }
 
 uint32
 wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
-                         uint64 expect, int64 timeout, bool wait64)
+                         uint64 expect, int64 timeout, bool wait64,
+                         WASMExecEnv *exec_env)
```

Comment:
Had better not add the new argument since the API is also called from AOT code, we can use `exec_env = wasm_clusters_search_exec_env((WASMModuleInstanceCommon *)module_inst)`. If the performance is not so good, we may consider improving it future by saving the `exec_env` into module_inst.

Reply:
Fixed